README.md
Split It!<img width="14%" align="right" src=".graphics/splitit-icon.svg">
=========

This is a simple program to reformat certain spreadsheets of results downloaded by Caltech Library staff from caltech.tind.io when they are doing inventory.

[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg?style=flat-square)](https://choosealicense.com/licenses/bsd-3-clause)
[![Latest release](https://img.shields.io/badge/Latest_release-1.1.0-b44e88.svg?style=flat-square)](http://shields.io)


Table of contents
-----------------

* [Introduction](#introduction)
* [Installation](#installation)
* [Usage](#usage)
* [Known issues and limitations](#known-issues-and-limitations)
* [Getting help](#getting-help)
* [Contributing](#contributing)
* [License](#license)
* [Authors and history](#authors-and-history)
* [Acknowledgments](#authors-and-acknowledgments)


Introduction
------------

The librarians at Caltech Library periodically have to perform library inventory.  In the process of doing that, they download spreadsheets of data from caltech.tind.io that have a format illustrated by the following example fragment:

```
574524,35047011136967,on shelf,,QA7 .A664 1991,
501345,350470002009169; 35047010046266,on shelf; on shelf,,QA7 .A67 1983,
381367,350470000767183; 350470000767192; 35047010794485,on shelf; on shelf; Limited circulation,,QA7 .S44,
```

The 2nd and 3rd lines above show examples where there are multiple results in a row, separated by semicolon (`;`) characters.  In these rows, columns 2 and 3 are parallel mappings of values, meaning that the barcode numbers in column 2 should be matched to the corresponding values in the semicolon-separated list of the 3rd column.  The other column values apply to each of the individual values in the compound.

_Split It!_ is a simple program that takes such as spreadsheet, splits the compound rows into separate rows, and produces a new spreadsheet with the collected results.  For the fragment above, it would look like this:

```
574524,35047011136967,on shelf,,QA7 .A664 1991,
501345,350470002009169,on shelf,,QA7 .A67 1983,
501345,35047010046266,on shelf,,QA7 .A67 1983,
381367,350470000767183,on shelf,,QA7 .S44,
381367,350470000767192,on shelf,,QA7 .S44,
381367,35047010794485,Limited circulation,,QA7 .S44,
```

Installation
------------

Binary installers for Windows can be downloaded from [the project's releases page on GitHub](https://github.com/caltechlibrary/splitit/releases).  Alternatively, you can use Python `pip` to install this from the repository using the following command:
```sh
sudo python3 -m pip install git+https://github.com/caltechlibrary/splitit.git --upgrade
```

As a final alternative, you can instead clone this GitHub repository and then run `setup.py` manually.  First, create a directory somewhere on your computer where you want to store the files, and cd to it from a terminal shell.  Next, execute the following commands:
```sh
git clone https://github.com/caltechlibrary/splitit.git
cd splitit
sudo python3 -m pip install . --upgrade
```


Usage
-----

_Split It!_ is a command line program.  It can be used from a terminal shell.  On all systems, the installation _should_ place a new program on your shell's search path called `splitit` (or `splitit.exe` on Windows), so that you can start _Split It!_ with a simple terminal command:
```csh
splitit
```

_Split It!_ accepts various command-line arguments.  To get information about the available options, use the `-h` argument (or `/h` on Windows):
```csh
splitit -h
```

In normal operation, _Split It!_ does not need to be started with any options; it will use GUI file dialogs to prompt for a file to open as input and a destination where to write the output.  Alternatively, these files can be provided on the command line via two options: an `-i` option (`/i` on Windows) to indicate the input CSV file, and an `-o` option (`/o` on Windows) to indicate the file where it should write the output in CSV format.  Here's an example to illustrate this:
```csh
splitit -i input.csv -o output.csv
```

If one or the other are not supplied, _Split It!_ will resort to using GUI file dialogs, unless the option `-G` (`/G` on Windows) is used to indicate that no GUI should be used.


Known issues and limitations
----------------------------

_Split It!_ currently assumes that the input spreadsheet has a format consisting of 5 columns, with the 2nd and 3rd columns being the ones that contain semicolon-separated values.  It does not verify that the input spreadsheet has this format; it simply proceeds on that assumption.  If the input spreadsheet does not conform to this format, the results are unpredictable.


Getting help
------------

If you find an issue, please submit it in [the GitHub issue tracker](https://github.com/caltechlibrary/splitit/issues) for this repository.


Contributing
------------

We would be happy to receive your help and participation with enhancing Split It!  Please visit the [guidelines for contributing](CONTRIBUTING.md) for some tips on getting started.


License
-------

Software produced by the Caltech Library is Copyright (C) 2019, Caltech.  This software is freely distributed under a BSD/MIT type license.  Please see the [LICENSE](LICENSE) file for more information.


Authors and history
---------------------------

[Mike Hucka](https://github.com/mhucka) designed and implemented Split It! based on requests from [Laurel Narizny](https://github.com/lnarizny) in mid-2019.


Acknowledgments
---------------

This work was funded by the California Institute of Technology Library.

The [vector artwork](https://thenounproject.com/term/page-break/31219/) of a "page break" icon used as a starting point for the logo for this repository was created by [Garrett Knoll](https://thenounproject.com/g_a.k_/) for the [Noun Project](https://thenounproject.com).  It is licensed under the Creative Commons [Attribution 3.0 Unported](https://creativecommons.org/licenses/by/3.0/deed.en) license.  The vector graphics was modified by Mike Hucka to change the color. 

_Split It!_ makes use of numerous open-source packages, without which it would have been effectively impossible to develop _Split It!_.  I want to acknowledge this debt.  In alphabetical order, the packages are:

* [colorama](https://github.com/tartley/colorama) &ndash; makes ANSI escape character sequences work under MS Windows terminals
* [ipdb](https://github.com/gotcha/ipdb) &ndash; the IPython debugger
* [plac](http://micheles.github.io/plac/) &ndash; a command line argument parser
* [setuptools](https://github.com/pypa/setuptools) &ndash; library for `setup.py`
* [termcolor](https://pypi.org/project/termcolor/) &ndash; ANSI color formatting for output in terminal
* [PyInstaller](http://www.pyinstaller.org) &ndash; a packaging program that creates standalone applications from Python programs for Windows, macOS, Linux and other platforms


<div align="center">
  <br>
  <a href="https://www.caltech.edu">
    <img width="100" height="100" src=".graphics/caltech-round.svg">
  </a>
</div>


CHANGES.md
Change log for Split It!
========================

Version 1.0.0
-------------

* Initial release



splitit/__main__.py
'''Split It!: Reformat spreadsheet of inventory results from Caltech.tind.io

Authors
-------

Michael Hucka <mhucka@caltech.edu> -- Caltech Library

Copyright
---------

Copyright (c) 2019 by the California Institute of Technology.  This code is
open-source software released under a 3-clause BSD license.  Please see the
file "LICENSE" for more information.
'''

import csv
import os
from   os import path
import plac
import sys
from   sys import exit as exit

import splitit
from splitit.debug import set_debug, log
from splitit.exceptions import *
from splitit.files import readable, writable, file_in_use, is_csv
from splitit.files import file_to_open, file_to_save
from splitit.messages import MessageHandlerCLI


# Main program.
# .............................................................................

@plac.annotations(
    no_gui     = ('do not use GUI dialogs to ask for files (default: do)', 'flag',   'G'),
    input_csv  = ('input file to be reformatted',                          'option', 'i'),
    output_csv = ('output file where results should be written',           'option', 'o'),
    no_color   = ('do not color-code terminal output',                     'flag',   'C'),
    quiet      = ('only print important messages while working',           'flag',   'q'),
    version    = ('print version info and exit',                           'flag',   'V'),
    debug      = ('turn on debug tracing & exception catching',            'flag',   '@'),
)

def main(no_gui = False, input_csv = 'I', output_csv = 'O', no_color = False,
         quiet = False, version = False, debug = False):
    '''Split It!

If the options -i and/or -o (or /i and /o on Windows) are not supplied, this
program will use GUI file dialogs to ask the user for the input and/or output
files (respectively) unless the option -G (/G on Windows) is used.

If the -G option (/G on Windows) is supplied to prevent the use of the GUI,
then this program must be invoked with two command-line options and values:
-i and -o (or /i and /o on Windows).  The -i option (/i on Windows) should be
followed by the path to an input file in CSV format that contains the content
to be reformatted; the -o option (/o on Windows) should be followed by the
path to a new, reformatted CSV file that should be written with the output.
Here is an example:

  splitit -i downloaded.csv -o inventory.csv

Additional command-line arguments
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If given the -q option (/q on Windows), this program will not print its usual
informational messages while it is working.  It will only print messages for
warnings or errors.  By default, messages are also color-coded.  If given the
option -C (/C on Windows), this program will not color the text of messages
it prints.  (This latter option is useful when running the program within
subshells inside other environments such as Emacs.)

If given the -V option (/V on Windows), this program will print the version
and other information, and exit without doing anything else.

If given the -@ option (/@ on Windows), this program will print additional
diagnostic output as it runs; in addition, it will start the Python debugger
(pdb) when an exception occurs, instead of simply exiting.

Command-line arguments summary
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
'''

    # Initial setup -----------------------------------------------------------

    say = MessageHandlerCLI(not no_color, quiet)
    prefix = '/' if sys.platform.startswith('win') else '-'
    hint = '(Hint: use {}h for help.)'.format(prefix)
    use_gui = not no_gui

    # Preprocess arguments and handle early exits -----------------------------

    if debug:
        set_debug(True)
    if version:
        print_version()
        exit()

    if input_csv == 'I' and use_gui:
        input_csv = file_to_open(splitit.__title__ + ': open input CSV file',
                                 wildcard = 'CSV file (*.csv)|*.csv|Any file (*.*)|*.*')
        if input_csv is None:
            exit('Quitting.')
    else:
        exit(say.error_text('Must supply input file using -i. {}'.format(hint)))
    if not readable(input_csv):
        exit(say.error_text('Cannot read file: {}'.format(input_csv)))
    elif not is_csv(input_csv):
        exit(say.error_text('File does not appear to contain CSV: {}'.format(input_csv)))

    if output_csv == 'O' and use_gui:
        output_csv = file_to_save(splitit.__title__ + ': save output file')
        if output_csv is None:
            exit('Quitting.')
    else:
        exit(say.error_text('Must supply output file using -o. {}'.format(hint)))
    if path.exists(output_csv):
        if file_in_use(output_csv):
            exit(say.error_text('File is open by another application: {}'.format(output_csv)))
        elif not writable(output_csv):
            exit(say.error_text('Unable to write to file: {}'.format(output_csv)))
    else:
        dest_dir = path.dirname(output_csv) or os.getcwd()
        if not writable(dest_dir):
            exit(say.error_text('Cannot write to folder: {}'.format(dest_dir)))

    # Do the real work --------------------------------------------------------

    # Example of possible input:
    #
    # 574524,35047011136967,on shelf,,QA7 .A664 1991,
    # 501345,350470002009169; 35047010046266,on shelf; on shelf,,QA7 .A67 1983,

    try:
        say.info('┏━━━━━━━━━━━━━━━━━┓')
        say.info('┃    Split It!    ┃')
        say.info('┗━━━━━━━━━━━━━━━━━┛')

        # Read it, massage it, write it.
        say.info('Reading input from "{}"'.format(input_csv))
        input_rows = []
        with open(input_csv, newline = '', encoding = 'utf8') as csvfile:
            input_rows = list(csv.reader(csvfile))

        # Simple-minded approach to splitting compound results
        output_rows = []
        for row in input_rows:
            if not row or row[0] == '':
                continue
            if ';' in row[1]:
                new_rows = []
                for index, part in enumerate(row[1].split(';')):
                    new_rows.append([row[0], part.strip()])
                for index, part in enumerate(row[2].split(';')):
                    new_rows[index].append(part.strip())
                for index in range(0, len(new_rows)):
                    new_rows[index].append(row[3].strip())
                    new_rows[index].append(row[4].strip())
                    new_rows[index].append(row[5].strip())
                output_rows += new_rows
            else:
                output_rows.append(row)

        say.info('Writing to "{}"'.format(output_csv))
        with open(output_csv, 'w', encoding = 'utf8') as csvfile:
            wr = csv.writer(csvfile, lineterminator = '\n')
            wr.writerows(output_rows)
    except (KeyboardInterrupt, UserCancelled) as ex:
        if __debug__: log('received {}', ex.__name__)
        exit(say.info_text('Quitting.'))
    except Exception as ex:
        if debug:
            import traceback
            say.error('{}\n{}'.format(str(ex), traceback.format_exc()))
            import pdb; pdb.set_trace()
        else:
            exit(say.error_text(str(ex)))
    say.info('Done.')


# Helper functions.
# .............................................................................

def print_version():
    print('{} version {}'.format(splitit.__title__, splitit.__version__))
    print('Author: {}'.format(splitit.__author__))
    print('URL: {}'.format(splitit.__url__))
    print('License: {}'.format(splitit.__license__))


# Main entry point.
# .............................................................................

# On windows, we want plac to use slash intead of hyphen for cmd-line options.
if sys.platform.startswith('win'):
    main.prefix_chars = '/'

# The following allows users to invoke this using "python3 -m splitit".
if __name__ == '__main__':
    plac.call(main)
